crates/circuit/src/circuit_data.rs

```diff
@@ -1105,4 +1024,213 @@ impl CircuitData {
+                                        expr.bind_borrowed(py),
+                                        &param_ob,
+                                        &value,
+                                        false,
+                                    )?;
+                                    // Historically, `assign_parameters` called `validate_parameter`
+                                    // only when a `ParameterExpression` became fully bound.  Some
+                                    // "generalised" (or user) gates fail without this, though
+                                    // arguably, that's them indicating they shouldn't be allowed to
+                                    // be parametric.
+                                    //
+                                    // Our `bind_expr` coercion means that a non-parametric
+                                    // `ParameterExperssion` after binding would have been coerced
+                                    // to a numeric quantity already, so the match here is
+                                    // definitely parameterized.
+                                    match new_param {
+                                        Param::ParameterExpression(_) => new_param,
+                                        new_param => Param::extract_no_coerce(&op.call_method1(
+                                            validate_parameter_attr,
+                                            (new_param,),
+                                        )?)?,
+                                    }
+                                }
+                                Param::Obj(obj) => {
+                                    let obj = obj.bind_borrowed(py);
+                                    if !obj.is_instance(QUANTUM_CIRCUIT.get_bound(py))? {
+                                        return Err(inconsistent());
+                                    }
+                                    Param::extract_no_coerce(
+                                        &obj.call_method(
+                                            assign_parameters_attr,
+                                            ([(&param_ob, &value)].into_py_dict_bound(py),),
+                                            Some(
+                                                &[("inplace", false), ("flat_input", true)]
+                                                    .into_py_dict_bound(py),
+                                            ),
+                                        )?,
+                                    )?
+                                }
+                            };
+                            op.getattr(params_attr)?.set_item(parameter, new_param)?;
+                            let mut new_op = op.extract::<OperationFromPython>()?;
+                            previous.op = new_op.operation;
+                            previous.params_mut().swap_with_slice(&mut new_op.params);
```

Comment:
I didn't know about this method, I like it. It's uses are a bit niche, but when you need it like here it is handy.

Reply:
Yeah, I learned about when writing this as well.  It _felt_ like a method that would exist in Rust, so I had a quick look.